src/PostgREST/Auth.hs

```diff
@@ -107,16 +106,49 @@ middleware appState app req respond = do
   let token  = fromMaybe "" $ Wai.extractBearerAuth =<< lookup HTTP.hAuthorization (Wai.requestHeaders req)
       parseJwt = runExceptT $ parseToken conf (LBS.fromStrict token) time >>= parseClaims conf
 
-  if configDbPlanEnabled conf
-    then do
-      (dur,authResult) <- timeItT parseJwt
-      let req' = req { Wai.vault = Wai.vault req & Vault.insert authResultKey authResult & Vault.insert jwtDurKey dur }
-      app req' respond
-    else do
-      authResult <- parseJwt
-      let req' = req { Wai.vault = Wai.vault req & Vault.insert authResultKey authResult }
-      app req' respond
-
+-- If DbPlanEnabled -> calculate JWT validation time
+-- If JwtCaching    -> cache JWT validation result
+  case (configDbPlanEnabled conf, configJwtCaching conf) of
+    (True, True)   -> do
+          (dur, authResult) <- timeItT $ getJWTFromCache appState token parseJwt
+          let req' = req { Wai.vault = Wai.vault req & Vault.insert authResultKey authResult & Vault.insert jwtDurKey dur }
+          app req' respond
+
+    (True, False)  -> do
+          (dur, authResult) <- timeItT parseJwt
+          let req' = req { Wai.vault = Wai.vault req & Vault.insert authResultKey authResult & Vault.insert jwtDurKey dur }
+          app req' respond
+
+    (False, True)  -> do
+          authResult <- getJWTFromCache appState token parseJwt
+          let req' = req { Wai.vault = Wai.vault req & Vault.insert authResultKey authResult }
+          app req' respond
+
+    (False, False) -> do
+          authResult <- parseJwt
+          let req' = req { Wai.vault = Wai.vault req & Vault.insert authResultKey authResult }
+          app req' respond
+
+-- Used to extract JWT exp claim and add to JWT Cache
+getTimeSpec :: AuthResult -> Maybe TimeSpec
+getTimeSpec res = do
+  let sciToInt = fromMaybe 0 . Sci.toBoundedInteger
+  expireJSON <- KM.lookup "exp" (authClaims res)
+  case expireJSON of
+    JSON.Number seconds -> Just $ TimeSpec (sciToInt seconds) 0
+    _                   -> Just $ TimeSpec 0 0 -- set timeSpec to 0 so it expires immediately, hence not cached
```

Comment:
The `0` here means that when there's no exp claim, the JWT will not be cached right?

How about an io test that confirms this. I guess the test would be that the duration is not down by 200 on the second request?

This would appease codecov here too.

Reply:
> If there is no exp claim, a JWT is valid infinitely, iirc. Imho, it should be chached without expiry, too.

Hm, one advantage of only caching JWTs with `exp` claims is that the memory consumption will always be bounded.

If a server gets too many infinitely valid JWTs, we could hit OOM errors and then crash.

How should we deal with this case? Should we force a cache lifetime for infinitely valid JWTs?

Reply:
Cool, agreed. 

>  jwt-cache-for = seconds

An `exp` might have a shorter time than the config, thus last less in the cache. So I think a more apt name would be: `jwt-cache-max-lifetime`. `0` for disabling would still stand.

Reply:
Just to be clear, is the `0` here relative to current time? I am not sure because `exp` is relative to the `POSIX` time AFAIK.

Reply:
Hm, no. Should just be a time in seconds for the cache max lifetime. So a lifetime of 24 hours would be:

```
jwt-cache-max-lifetime = 86400
```